Fix #4089 Linked account cancel

kevcunnane opened the pull request:
VSCode serialization changed rejection handling to only serialize errors, which caused things to break
- Changed to return either account info or a cancel message in the resolve
- Rewrote to use promises. Tracking how to return canceled through 4+ thenables was way trickier than just using a promise
- Updated unit tests to handle new scenario
- Tested integration tests, realized they a) didn't run and b) didn't passed. 
  - Added vscode dev dependency to fix run issue
  - Fixed tests to account for behavior changes in tree state.

anthonydresser commented on extensions/azurecore/package.json:
```diff
@@ -169,6 +169,7 @@
 		"should": "^13.2.1",
-		"typemoq": "^2.1.0"
+		"typemoq": "^2.1.0",
+		"vscode": "^1.1.33"
```
You shouldn't need to bring this in here. we ref the .d.ts directly https://github.com/Microsoft/azuredatastudio/blob/master/extensions/azurecore/src/typings/ref.d.ts#L6

kevcunnane replied:
I need this for the bin folder - testrunner is referenced from `const testRunner = require('vscode/lib/testrunner');` in the main index.ts. Without this, integration tests were broken.

Incidentally we might want to look at an alternate way for our built-in extensions to reference and use this, but this is how a few others already work.

kevcunnane commented:
@anthonydresser any issues? I addressed the 1 comment here, and fixed build break caused by vscode.d.ts conflicts. I didn't see an approved / request changes from you?

anthonydresser commented on extensions/azurecore/package.json:
```diff
@@ -169,6 +169,7 @@
 		"should": "^13.2.1",
-		"typemoq": "^2.1.0"
+		"typemoq": "^2.1.0",
+		"vscode": "1.1.26"
```
You still shouldn't need this, like I said we are directly refering to the .d.ts in our src.

anthonydresser replied:
I just saw your previous comment

anthonydresser replied:
Presumably there is a way to run integration tests without the need to refer to vscode npm module, since git extension doesn't and it has colorized tests which run as integration tests, but maybe thats an issue for someone else to deal with...

kevcunnane replied:
I looked at Git. What happens is extensions/vscode-colorize-tests/package.json has a VSCode dependency, and it finds all extensions with a `test/colorize-fixtures` folder and uses those in its test framework. So, even the built-in VSCode tests use this pattern.

anthonydresser commented on extensions/azurecore/tsconfig.json:
```diff
@@ -1,14 +1,22 @@
 {
-	"extends": "../shared.tsconfig.json",
+	"compileOnSave": true,
```
why are you reverting this?

anthonydresser replied:
Is this required from requiring vscode directly? would just removing the .d.ts reference in the code fix compiling without needing this reverted?

anthonydresser replied:
Importantly, the shared tsconfig contains a bunch of strict compile options that help ensure strict (good practice) coding guidelines, so by revert this, we are no longer enforcing those.

kevcunnane replied:
Thanks - I had been trying to fix the failing build and tried changing (again, to another of our extensions) .tsconfig settings that I knew a) used VSCode module and b) did compile. Having reverted the file, it looks like it still works. So it's a problem with newer VSCode module versions where they always bundle in a vscode.d.ts file, causing duplication problems. 

Anyhow, this should be reverted in the next iteration

kevcunnane commented:
@anthonydresser all comments should be resolved. I didn't see you request changes - are there any blockers to signoff here?
